### PR TITLE
Writing docs for the new doc-builder

### PR DESCRIPTION
**Pull request by KareemMAX**
The files I changed are accessible here:
- https://kareemmax.github.io/ivy/overview/contributing/building_the_docs.html
- https://kareemmax.github.io/ivy/overview/deep_dive/building_the_docs_pipline.html
- https://kareemmax.github.io/ivy/overview/deep_dive.html (Added u)

### CODE REVIEW

File: docs/overview/deep_dive/building_the_docs_pipline.rst [hunk 1]
```diff
@@ -0,0 +1,482 @@
...
+4. It can contain an optional ``docs/prebuild.sh`` file, which will be executed before
+   the docs are built. This is useful if you need to install some dependencies for the
+   docs to build.
+
+5. It can contain an optional ``docs/partial_conf.py`` which is a partial `Sphinx
+   configuration file`_.
+   This file will be imported with the default ``conf.py`` file located in the 
+   ``doc-builder`` repo.
+
+Running the script like this:
```
Comment (CatB1t) [inline docs/overview/deep_dive/building_the_docs_pipline.rst:48]: Let's omit "like this"

File: docs/overview/deep_dive/building_the_docs_pipline.rst [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+
+.. code-block:: bash
+
+    ./make_docs_without_docker.sh /path/to/project
+
+will build the docs for the project and output them in the ``docs/build`` folder.
```
Comment (CatB1t) [inline docs/overview/deep_dive/building_the_docs_pipline.rst:54]: will result in the creation of documentation for the project in the directory ``doc/build``

File: docs/overview/deep_dive/building_the_docs_pipline.rst [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+
+The Docker image
+~~~~~~~~~~~~~~~~
+
+The Docker image `unifyai/doc-builder <https://hub.docker.com/r/unifyai/doc-builder>`_
+works as a wrapper around the ``make_docs_without_docker.sh`` script. It basically run
```
Comment (CatB1t) [inline docs/overview/deep_dive/building_the_docs_pipline.rst:60]: Let's omit "basically", I think 'run' should be 'runs'

File: docs/overview/deep_dive/building_the_docs_pipline.rst [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+the script on the ``/project`` folder, located in the container `like this
```
Comment (CatB1t) [inline docs/overview/deep_dive/building_the_docs_pipline.rst:61]: replace 'folder' with 'directory', replace 'like this' with 'as shown here'

File: docs/overview/deep_dive/building_the_docs_pipline.rst [hunk 5]
```diff
@@ -0,0 +1,482 @@
...
+The Docker image `unifyai/doc-builder <https://hub.docker.com/r/unifyai/doc-builder>`_
+works as a wrapper around the ``make_docs_without_docker.sh`` script. It basically run
+the script on the ``/project`` folder, located in the container `like this 
+<https://github.com/unifyai/doc-builder/blob/master/Dockerfile#L20>`_:
+
+.. code-block:: bash
+
+    ./make_docs_without_docker.sh /project
+
+So, to build the docs through docker you use this command:
```
Comment (CatB1t) [inline docs/overview/deep_dive/building_the_docs_pipline.rst:68]: No need for 'So,' here.

File: docs/overview/deep_dive/building_the_docs_pipline.rst [hunk 6]
```diff
@@ -0,0 +1,482 @@
...
+
+.. code-block:: bash
+
+    docker run -v /path/to/project:/project unifyai/doc-builder
+
+How Ivy's docs is structured
+-----------------------------
+
+If you looked into `Ivy's docs <https://github.com/unifyai/ivy/tree/master/docs>`_,
```
Comment (CatB1t) [inline docs/overview/deep_dive/building_the_docs_pipline.rst:77]: Let's rephrase this to "Looking at Ivy docs, we can see that..."

File: docs/overview/deep_dive/building_the_docs_pipline.rst [hunk 7]
```diff
@@ -0,0 +1,482 @@
...
+        "ivy.functional.ivy": "Functions",
+        "ivy.stateful": "Framework classes",
+        "ivy.nested_array": "Nested array",
+        "ivy.utils": "Utils",
+        "ivy_tests.test_ivy.helpers": "Testing",
+    }
+
+Here we are overriding the ``ivy_toctree_caption_map`` configuration, which is used to 
+customize the title of the table of contents for each module. This is specific to the 
+custom ``autosummary`` extension and we will cover all configurations in
```
Comment (CatB1t) [inline docs/overview/deep_dive/building_the_docs_pipline.rst:175]: the second statement doesn't feel right, try to re-write it

File: docs/overview/deep_dive/building_the_docs_pipline.rst [hunk 8]
```diff
@@ -0,0 +1,482 @@
...
+
+``:fix-directory:``
+"""""""""""""""""""
+
+Because of the nature of the ``autosummary`` directive, it generates stub files relative
+to the current file. If we used include, and there is an ``autosummary`` directive in
+the stub file, this directive will become invalid, because sphinx include the stub file
+by substitution.
+
+Let's say you have a file called ``index.rst`` which has this in it:
```
Comment (CatB1t) [inline docs/overview/deep_dive/building_the_docs_pipline.rst:282]: Let's use "let's assume" instead of "let's say"

File: docs/overview/deep_dive/building_the_docs_pipline.rst [hunk 9]
```diff
@@ -0,0 +1,482 @@
...
+
+    .. autosummary::
+        :toctree: toctree
+        :template: top_level_toc.rst
+        :recursive:
+        :include:
+
+        module
+
+If we said that ``module`` have 2 submodules ``foo``, and ``bar``, then the generated
```
Comment (CatB1t) [inline docs/overview/deep_dive/building_the_docs_pipline.rst:294]: Use "let's assume" instead of "if we said"

File: docs/overview/deep_dive/building_the_docs_pipline.rst [hunk 10]
```diff
@@ -0,0 +1,482 @@
...
+        :template: top_level_module.rst
+        :caption: module
+        :substitute-caption:
+        :hide-table:
+        :fix-directory:
+
+        foo
+        bar
+
+You see? The ``:toctree:`` option is now invalid, because it's now pointing to the
```
Comment (CatB1t) [inline docs/overview/deep_dive/building_the_docs_pipline.rst:342]: let's omit "You see?"

**Review by CatB1t — CHANGES_REQUESTED** [review]

File: docs/overview/deep_dive.rst [hunk 11]
```diff
@@ -73,7 +73,10 @@ We're excited for you to get involved! 🦾
 | (t) :ref:`Continuous Integration` 🔁
 | Ivy Tests running on the Repository
 |
-| (u) :ref:`Building the Docs Pipeline` 📚
+| (u) :ref:`Gradients` 🔁
```
Comment (CatB1t) [inline docs/overview/deep_dive.rst:76]: Are these changes intentional?

File: docs/partial_conf.py [hunk 12]
```diff
@@ -48,3 +48,9 @@
 # That resolved a bug of autosummary generating docs for code-block examples
 # of autosummary
 autosummary_generate = ['index.rst']
+
+skippable_method_attributes = [
```
Comment (CatB1t) [inline docs/partial_conf.py:52]: Same for this one here.

File: docs/overview/deep_dive.rst [hunk 13]
```diff
@@ -100,4 +103,5 @@ We're excited for you to get involved! 🦾
    deep_dive/ivy_frontends_tests.rst
    deep_dive/exception_handling.rst
    deep_dive/continuous_integration.rst
+   deep_dive/gradients.rst
```
Comment (CatB1t) [inline docs/overview/deep_dive.rst:106]: This one too.

**Review by CatB1t — CHANGES_REQUESTED** [review]
Hello @KareemMAX, lgtm! just left a few points regarding the changes that seem not fully related to the title of the PR, it would be better if we can separate these changes from this PR so it's not confusing.

File: docs/partial_conf.py (hunk 12 above)
Comment (KareemMAX) [reply docs/partial_conf.py:52]: Yup, it's explained in the comment above it
